<a href="https://colab.research.google.com/github/ramzymhafeel/SE4050-Deep-Learning-Lab-01/blob/main/notebooks/SE4050_Lab01_Task2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SE4050 — Deep Learning

## Lab 01: Introduction to Python Libraries for Deep Learning

### Task 2 — Au Nanoparticle Dataset Analysis

**Student Name:** Mohamed Ramzy  
**Index Number:** IT23386822
**Module:** SE4050 — Deep Learning  
**Academic Year:** 2026

## Objective

The objective of Task 2 is to use Pandas, Matplotlib, and Seaborn to analyze and visualize selected features from the Au nanoparticle dataset.

The selected features are:

- `N_total`
- `N_bulk`
- `N_surface`
- `R_avg`

In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print("Required libraries imported successfully.")

Required libraries imported successfully.


In [2]:
from google.colab import drive

drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
dataset_folder = (
    "/content/drive/MyDrive/"
    "SE4050_DeepLearning/"
    "Lab_01/"
    "02_Datasets"
)

au_file_path = os.path.join(
    dataset_folder,
    "Au_nanoparticle_dataset.csv"
)

print("Dataset path:")
print(au_file_path)

Dataset path:
/content/drive/MyDrive/SE4050_DeepLearning/Lab_01/02_Datasets/Au_nanoparticle_dataset.csv


## Task 2.1 — Load the Au Nanoparticle Dataset

### Introduction

The Au nanoparticle dataset is loaded using Pandas for data analysis and visualization.

The analysis focuses on the following four numerical features:

- `N_total`
- `N_bulk`
- `N_surface`
- `R_avg`

In [4]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print("Required libraries imported successfully.")

Required libraries imported successfully.


In [5]:
from google.colab import drive

drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
dataset_folder = (
    "/content/drive/MyDrive/"
    "SE4050_DeepLearning/"
    "Lab_01/"
    "02_Datasets"
)

au_file_path = os.path.join(
    dataset_folder,
    "Au_nanoparticle_dataset.csv"
)

print("Dataset path:")
print(au_file_path)

Dataset path:
/content/drive/MyDrive/SE4050_DeepLearning/Lab_01/02_Datasets/Au_nanoparticle_dataset.csv


In [7]:
if os.path.exists(au_file_path):
    print("Au nanoparticle dataset found successfully.")
else:
    print("Dataset file was not found.")

Au nanoparticle dataset found successfully.


In [8]:
au_df = pd.read_csv(au_file_path)

print("Dataset loaded successfully.")
print("Dataset shape:", au_df.shape)

Dataset loaded successfully.
Dataset shape: (4000, 185)


In [9]:
au_df.head()

,ID,T,tau,time,N_total,N_bulk,N_surface,Volume,R_min,R_max,...,q6q6_S14,q6q6_S15,q6q6_S16,q6q6_S17,q6q6_S18,q6q6_S19,q6q6_S20,q6q6_S20+,Total_E,Formation_E
0,1,273,0.000025,1,1599,1014,585,2.710000e-26,15.5898,19.4516,...,0,0,0,0,0,0,0,0,-5908.6345,343.4555
1,2,273,0.000025,2,1642,1034,608,2.780000e-26,15.3996,24.1755,...,0,0,0,0,0,0,0,0,-6035.4034,384.8166
2,3,273,0.000025,3,4637,3365,1272,7.850000e-26,21.9199,31.0305,...,0,0,0,0,0,0,0,0,-17330.1660,800.5040
3,4,273,0.000025,4,7189,5292,1897,1.220000e-25,23.2542,37.8795,...,0,0,0,0,0,0,0,0,-26937.2010,1171.7890
4,5,273,0.000025,5,11004,8508,2496,1.860000e-25,27.1200,39.7614,...,0,0,0,0,0,0,0,0,-41427.9010,1597.7390


In [10]:
au_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Columns: 185 entries, ID to Formation_E
dtypes: float64(23), int64(162)
memory usage: 5.6 MB


In [11]:
print("Column names in the dataset:\n")

for number, column in enumerate(
    au_df.columns,
    start=1
):
    print(f"{number}. {repr(column)}")

Column names in the dataset:

1. 'ID'
2. 'T'
3. 'tau'
4. 'time'
5. 'N_total'
6. 'N_bulk'
7. 'N_surface'
8. 'Volume'
9. 'R_min'
10. 'R_max'
11. 'R_diff'
12. 'R_avg'
13. 'R_std'
14. 'R_skew'
15. 'R_kurt'
16. 'S_100'
17. 'S_111'
18. 'S_110'
19. 'S_311'
20. 'Curve_1-10'
21. 'Curve_11-20'
22. 'Curve_21-30'
23. 'Curve_31-40'
24. 'Curve_41-50'
25. 'Curve_51-60'
26. 'Curve_61-70'
27. 'Curve_71-80'
28. 'Curve_81-90'
29. 'Curve_91-100'
30. 'Curve_101-110'
31. 'Curve_111-120'
32. 'Curve_121-130'
33. 'Curve_131-140'
34. 'Curve_141-150'
35. 'Curve_151-160'
36. 'Curve_161-170'
37. 'Curve_171-180'
38. 'Avg_total'
39. 'Avg_bulk'
40. 'Avg_surf'
41. 'TCN_0'
42. 'TCN_1'
43. 'TCN_2'
44. 'TCN_3'
45. 'TCN_4'
46. 'TCN_5'
47. 'TCN_6'
48. 'TCN_7'
49. 'TCN_8'
50. 'TCN_9'
51. 'TCN_10'
52. 'TCN_11'
53. 'TCN_12'
54. 'TCN_13'
55. 'TCN_14'
56. 'TCN_15'
57. 'TCN_16'
58. 'TCN_17'
59. 'TCN_18'
60. 'TCN_19'
61. 'TCN_20'
62. 'BCN_0'
63. 'BCN_1'
64. 'BCN_2'
65. 'BCN_3'
66. 'BCN_4'
67. 'BCN_5'
68. 'BCN_6'
69. 'BCN_7'
70. '

In [12]:
au_df.columns = au_df.columns.str.strip()

print("Column names cleaned successfully.")

Column names cleaned successfully.


In [13]:
required_columns = [
    "N_total",
    "N_bulk",
    "N_surface",
    "R_avg"
]

missing_columns = [
    column
    for column in required_columns
    if column not in au_df.columns
]

if not missing_columns:
    print("All four required columns are available.")

    for column in required_columns:
        print("✓", column)
else:
    print("Missing columns:", missing_columns)

All four required columns are available.
✓ N_total
✓ N_bulk
✓ N_surface
✓ R_avg


## Task 2.2 — Select the Required Features

### Feature Selection

A new dataframe is created using only the four features required by the lab sheet:

- `N_total`
- `N_bulk`
- `N_surface`
- `R_avg`

The original dataframe remains unchanged.

In [14]:
new_df = au_df[
    [
        "N_total",
        "N_bulk",
        "N_surface",
        "R_avg"
    ]
].copy()

print("Filtered dataframe created successfully.")
print("Filtered dataframe shape:", new_df.shape)

Filtered dataframe created successfully.
Filtered dataframe shape: (4000, 4)


In [15]:
print("Filtered dataframe columns:")
print(new_df.columns.tolist())

Filtered dataframe columns:
['N_total', 'N_bulk', 'N_surface', 'R_avg']


In [16]:
print("Data types of selected features:\n")
print(new_df.dtypes)

Data types of selected features:

N_total        int64
N_bulk         int64
N_surface      int64
R_avg        float64
dtype: object


In [17]:
for column in required_columns:
    new_df[column] = pd.to_numeric(
        new_df[column],
        errors="coerce"
    )

print("Selected features converted to numeric format.")

Selected features converted to numeric format.


In [18]:
print("Missing values in selected features:\n")
print(new_df.isnull().sum())

Missing values in selected features:

N_total      0
N_bulk       0
N_surface    0
R_avg        0
dtype: int64


## Task 2.3 — Display the First 20 Samples

### First 20 Samples

The first 20 samples of the filtered dataframe are displayed below using the Pandas `head()` function.

In [19]:
first_20_samples = new_df.head(20)

print("First 20 samples of the selected features:")

display(first_20_samples)

First 20 samples of the selected features:


,N_total,N_bulk,N_surface,R_avg
0,1599,1014,585,17.3706
1,1642,1034,608,17.6061
2,4637,3365,1272,25.3692
3,7189,5292,1897,29.7011
4,11004,8508,2496,34.2831
5,13375,10768,2607,36.6334
6,13795,11155,2640,37.0108
7,13947,11304,2643,37.1672
8,14020,11357,2663,37.2103
9,14056,11389,2667,37.2467


In [20]:
first_20_display = (
    new_df
    .head(20)
    .copy()
)

first_20_display.index = range(
    1,
    len(first_20_display) + 1
)

first_20_display.index.name = "Sample Number"

display(first_20_display)

,N_total,N_bulk,N_surface,R_avg
Sample Number,,,,
1,1599,1014,585,17.3706
2,1642,1034,608,17.6061
3,4637,3365,1272,25.3692
4,7189,5292,1897,29.7011
5,11004,8508,2496,34.2831
6,13375,10768,2607,36.6334
7,13795,11155,2640,37.0108
8,13947,11304,2643,37.1672
9,14020,11357,2663,37.2103


In [21]:
print(
    "Number of displayed samples:",
    first_20_samples.shape[0]
)

print(
    "Number of displayed features:",
    first_20_samples.shape[1]
)

Number of displayed samples: 20
Number of displayed features: 4


In [22]:
print("Original dataframe shape:", au_df.shape)
print("Filtered dataframe shape:", new_df.shape)

print(
    "Original number of columns:",
    au_df.shape[1]
)

print(
    "Filtered number of columns:",
    new_df.shape[1]
)

Original dataframe shape: (4000, 185)
Filtered dataframe shape: (4000, 4)
Original number of columns: 185
Filtered number of columns: 4


### Interpretation

The Au nanoparticle dataset was loaded successfully using Pandas.

A new dataframe named `new_df` was created containing only the four required features: `N_total`, `N_bulk`, `N_surface`, and `R_avg`.

The filtered dataframe contains four columns, and its first 20 samples were displayed successfully. The original dataset remains unchanged.

In [23]:
# Task 2.1 to Task 2.3:
# Load the Au nanoparticle dataset,
# select the required features,
# and display the first 20 samples.

import os
import pandas as pd

dataset_folder = (
    "/content/drive/MyDrive/"
    "SE4050_DeepLearning/"
    "Lab_01/"
    "02_Datasets"
)

au_file_path = os.path.join(
    dataset_folder,
    "Au_nanoparticle_dataset.csv"
)

au_df = pd.read_csv(au_file_path)
au_df.columns = au_df.columns.str.strip()

selected_features = [
    "N_total",
    "N_bulk",
    "N_surface",
    "R_avg"
]

new_df = au_df[selected_features].copy()

for column in selected_features:
    new_df[column] = pd.to_numeric(
        new_df[column],
        errors="coerce"
    )

print("Original dataset shape:", au_df.shape)
print("Filtered dataframe shape:", new_df.shape)
print("\nFirst 20 samples:")

display(new_df.head(20))

Original dataset shape: (4000, 185)
Filtered dataframe shape: (4000, 4)

First 20 samples:


,N_total,N_bulk,N_surface,R_avg
0,1599,1014,585,17.3706
1,1642,1034,608,17.6061
2,4637,3365,1272,25.3692
3,7189,5292,1897,29.7011
4,11004,8508,2496,34.2831
5,13375,10768,2607,36.6334
6,13795,11155,2640,37.0108
7,13947,11304,2643,37.1672
8,14020,11357,2663,37.2103
9,14056,11389,2667,37.2467


In [24]:
assert new_df.shape[1] == 4, (
    "The filtered dataframe must contain exactly four columns."
)

assert list(new_df.columns) == required_columns, (
    "The selected columns are incorrect or in the wrong order."
)

assert len(new_df.head(20)) == 20, (
    "The output must contain exactly 20 samples."
)

print("All Phase 6 validation tests passed.")

All Phase 6 validation tests passed.


## Task 2.4 — Descriptive Statistical Analysis

### Introduction

Descriptive statistics summarize the main characteristics of numerical data.

For the four selected Au nanoparticle features, the following measurements are calculated:

- Mean
- Standard deviation
- First quartile (`Q1`)
- Median (`Q2`)
- Third quartile (`Q3`)

Additional values such as the minimum, maximum, and interquartile range are also included to provide a clearer understanding of each feature.

In [25]:
print("Filtered dataframe shape:", new_df.shape)
print("Selected columns:", new_df.columns.tolist())

display(new_df.head())

Filtered dataframe shape: (4000, 4)
Selected columns: ['N_total', 'N_bulk', 'N_surface', 'R_avg']


,N_total,N_bulk,N_surface,R_avg
0,1599,1014,585,17.3706
1,1642,1034,608,17.6061
2,4637,3365,1272,25.3692
3,7189,5292,1897,29.7011
4,11004,8508,2496,34.2831


In [26]:
print("Missing values before statistical analysis:\n")
print(new_df.isnull().sum())

Missing values before statistical analysis:

N_total      0
N_bulk       0
N_surface    0
R_avg        0
dtype: int64


In [28]:
statistics_df = new_df.dropna().copy()

print("Rows before removing missing values:", new_df.shape[0])
print("Rows used for analysis:", statistics_df.shape[0])

Rows before removing missing values: 4000
Rows used for analysis: 4000


In [29]:
statistics_df = new_df.copy()

print("No rows were removed.")
print("Rows used for analysis:", statistics_df.shape[0])

No rows were removed.
Rows used for analysis: 4000


In [30]:
statistics_df = new_df.dropna().copy()

removed_rows = new_df.shape[0] - statistics_df.shape[0]

print("Original rows:", new_df.shape[0])
print("Rows used for analysis:", statistics_df.shape[0])
print("Rows removed due to missing values:", removed_rows)

Original rows: 4000
Rows used for analysis: 4000
Rows removed due to missing values: 0


In [31]:
feature_means = statistics_df.mean()

print("Mean values:\n")
print(feature_means)

Mean values:

N_total      3476.786500
N_bulk       2521.550250
N_surface     955.236250
R_avg          20.654363
dtype: float64


In [32]:
feature_standard_deviations = statistics_df.std()

print("Standard deviation values:\n")
print(feature_standard_deviations)

Standard deviation values:

N_total      3679.286769
N_bulk       2976.232459
N_surface     721.870220
R_avg           7.610716
dtype: float64


In [33]:
quartile_values = statistics_df.quantile(
    [0.25, 0.50, 0.75]
)

print("Quartile values:\n")
display(quartile_values)

Quartile values:



,N_total,N_bulk,N_surface,R_avg
0.25,1061.0,618.75,437.00,15.160725
0.50,1867.0,1199.00,666.00,18.629250
0.75,4503.0,3183.00,1301.75,25.525125


In [34]:
q1_values = statistics_df.quantile(0.25)
median_values = statistics_df.quantile(0.50)
q3_values = statistics_df.quantile(0.75)

print("First quartile, Q1:\n")
print(q1_values)

print("\nMedian, Q2:\n")
print(median_values)

print("\nThird quartile, Q3:\n")
print(q3_values)

First quartile, Q1:

N_total      1061.000000
N_bulk        618.750000
N_surface     437.000000
R_avg          15.160725
Name: 0.25, dtype: float64

Median, Q2:

N_total      1867.00000
N_bulk       1199.00000
N_surface     666.00000
R_avg          18.62925
Name: 0.5, dtype: float64

Third quartile, Q3:

N_total      4503.000000
N_bulk       3183.000000
N_surface    1301.750000
R_avg          25.525125
Name: 0.75, dtype: float64


In [35]:
iqr_values = q3_values - q1_values

print("Interquartile range values:\n")
print(iqr_values)

Interquartile range values:

N_total      3442.0000
N_bulk       2564.2500
N_surface     864.7500
R_avg          10.3644
dtype: float64


In [36]:
descriptive_summary = statistics_df.describe()

print("Complete descriptive statistical summary:")

display(descriptive_summary)

Complete descriptive statistical summary:


,N_total,N_bulk,N_surface,R_avg
count,4000.000000,4000.000000,4000.00000,4000.000000
mean,3476.786500,2521.550250,955.23625,20.654363
std,3679.286769,2976.232459,721.87022,7.610716
min,236.000000,89.000000,137.00000,8.528600
25%,1061.000000,618.750000,437.00000,15.160725
50%,1867.000000,1199.000000,666.00000,18.629250
75%,4503.000000,3183.000000,1301.75000,25.525125
max,14277.000000,11580.000000,3190.00000,39.239200


In [37]:
descriptive_summary_rounded = descriptive_summary.round(4)

display(descriptive_summary_rounded)

,N_total,N_bulk,N_surface,R_avg
count,4000.0000,4000.0000,4000.0000,4000.0000
mean,3476.7865,2521.5502,955.2362,20.6544
std,3679.2868,2976.2325,721.8702,7.6107
min,236.0000,89.0000,137.0000,8.5286
25%,1061.0000,618.7500,437.0000,15.1607
50%,1867.0000,1199.0000,666.0000,18.6292
75%,4503.0000,3183.0000,1301.7500,25.5251
max,14277.0000,11580.0000,3190.0000,39.2392


In [38]:
custom_statistics = pd.DataFrame({
    "Mean": statistics_df.mean(),
    "Standard Deviation": statistics_df.std(),
    "Minimum": statistics_df.min(),
    "Q1 (25%)": statistics_df.quantile(0.25),
    "Median (50%)": statistics_df.quantile(0.50),
    "Q3 (75%)": statistics_df.quantile(0.75),
    "Maximum": statistics_df.max()
})

custom_statistics["IQR"] = (
    custom_statistics["Q3 (75%)"]
    - custom_statistics["Q1 (25%)"]
)

print("Custom descriptive statistics table:")

display(custom_statistics.round(4))

Custom descriptive statistics table:


,Mean,Standard Deviation,Minimum,Q1 (25%),Median (50%),Q3 (75%),Maximum,IQR
N_total,3476.7865,3679.2868,236.0000,1061.0000,1867.0000,4503.0000,14277.0000,3442.0000
N_bulk,2521.5502,2976.2325,89.0000,618.7500,1199.0000,3183.0000,11580.0000,2564.2500
N_surface,955.2362,721.8702,137.0000,437.0000,666.0000,1301.7500,3190.0000,864.7500
R_avg,20.6544,7.6107,8.5286,15.1607,18.6292,25.5251,39.2392,10.3644


In [39]:
custom_statistics.index.name = "Feature"

display(custom_statistics.round(4))

,Mean,Standard Deviation,Minimum,Q1 (25%),Median (50%),Q3 (75%),Maximum,IQR
Feature,,,,,,,,
N_total,3476.7865,3679.2868,236.0000,1061.0000,1867.0000,4503.0000,14277.0000,3442.0000
N_bulk,2521.5502,2976.2325,89.0000,618.7500,1199.0000,3183.0000,11580.0000,2564.2500
N_surface,955.2362,721.8702,137.0000,437.0000,666.0000,1301.7500,3190.0000,864.7500
R_avg,20.6544,7.6107,8.5286,15.1607,18.6292,25.5251,39.2392,10.3644


In [40]:
n_total_mean = statistics_df["N_total"].mean()
n_total_std = statistics_df["N_total"].std()
n_total_q1 = statistics_df["N_total"].quantile(0.25)
n_total_median = statistics_df["N_total"].quantile(0.50)
n_total_q3 = statistics_df["N_total"].quantile(0.75)

print("Manual verification for N_total:")
print("Mean:", round(n_total_mean, 4))
print("Standard deviation:", round(n_total_std, 4))
print("Q1:", round(n_total_q1, 4))
print("Median:", round(n_total_median, 4))
print("Q3:", round(n_total_q3, 4))

Manual verification for N_total:
Mean: 3476.7865
Standard deviation: 3679.2868
Q1: 1061.0
Median: 1867.0
Q3: 4503.0


In [41]:
largest_mean_feature = feature_means.idxmax()
largest_mean_value = feature_means.max()

print("Feature with the largest mean:")
print(largest_mean_feature, "=", round(largest_mean_value, 4))

Feature with the largest mean:
N_total = 3476.7865


In [42]:
largest_std_feature = feature_standard_deviations.idxmax()
largest_std_value = feature_standard_deviations.max()

print("Feature with the largest standard deviation:")
print(largest_std_feature, "=", round(largest_std_value, 4))

Feature with the largest standard deviation:
N_total = 3679.2868


In [43]:
coefficient_of_variation = (
    statistics_df.std()
    / statistics_df.mean()
)

print("Coefficient of variation:\n")
print(coefficient_of_variation.round(4))

Coefficient of variation:

Feature
N_total      1.0582
N_bulk       1.1803
N_surface    0.7557
R_avg        0.3685
dtype: float64


### Interpretation

The descriptive statistics summarize the central tendency and variability of the four selected Au nanoparticle features.

- The **mean** represents the average value of each feature.
- The **standard deviation** indicates how widely values are distributed around the mean.
- The **first quartile (`Q1`)** represents the value below which approximately 25% of observations fall.
- The **median (`Q2`)** divides the data into two equal halves.
- The **third quartile (`Q3`)** represents the value below which approximately 75% of observations fall.
- The **interquartile range (`IQR`)** represents the spread of the middle 50% of observations.

Differences between the mean and median may indicate that a feature has a skewed distribution. A large standard deviation or IQR indicates that the values of that feature are more widely dispersed.

In [44]:
for column in statistics_df.columns:
    mean_value = statistics_df[column].mean()
    median_value = statistics_df[column].median()

    if mean_value > median_value:
        shape_message = "may be positively skewed"
    elif mean_value < median_value:
        shape_message = "may be negatively skewed"
    else:
        shape_message = "appears approximately symmetric"

    print(
        f"{column}: mean = {mean_value:.4f}, "
        f"median = {median_value:.4f}; "
        f"the distribution {shape_message}."
    )

N_total: mean = 3476.7865, median = 1867.0000; the distribution may be positively skewed.
N_bulk: mean = 2521.5502, median = 1199.0000; the distribution may be positively skewed.
N_surface: mean = 955.2363, median = 666.0000; the distribution may be positively skewed.
R_avg: mean = 20.6544, median = 18.6292; the distribution may be positively skewed.


### Dataset-Specific Observation

The mean and median values are not identical for all four features. This suggests that some of the distributions may be skewed.

The histogram analysis in the next section will be used to examine the shapes of these distributions visually.

In [45]:
# Task 2.4:
# Calculate mean, standard deviation, and quartile values.

statistics_df = new_df.dropna().copy()

statistics_table = pd.DataFrame({
    "Mean": statistics_df.mean(),
    "Standard Deviation": statistics_df.std(),
    "Q1 (25%)": statistics_df.quantile(0.25),
    "Median (50%)": statistics_df.quantile(0.50),
    "Q3 (75%)": statistics_df.quantile(0.75)
})

statistics_table["IQR"] = (
    statistics_table["Q3 (75%)"]
    - statistics_table["Q1 (25%)"]
)

statistics_table.index.name = "Feature"

print("Mean, standard deviation, and quartile values:")

display(statistics_table.round(4))

Mean, standard deviation, and quartile values:


,Mean,Standard Deviation,Q1 (25%),Median (50%),Q3 (75%),IQR
Feature,,,,,,
N_total,3476.7865,3679.2868,1061.0000,1867.0000,4503.0000,3442.0000
N_bulk,2521.5502,2976.2325,618.7500,1199.0000,3183.0000,2564.2500
N_surface,955.2362,721.8702,437.0000,666.0000,1301.7500,864.7500
R_avg,20.6544,7.6107,15.1607,18.6292,25.5251,10.3644


In [46]:
expected_features = [
    "N_total",
    "N_bulk",
    "N_surface",
    "R_avg"
]

assert list(statistics_df.columns) == expected_features, (
    "The statistical analysis must use the four required features."
)

assert statistics_table.shape[0] == 4, (
    "The statistics table must contain four feature rows."
)

assert "Mean" in statistics_table.columns
assert "Standard Deviation" in statistics_table.columns
assert "Q1 (25%)" in statistics_table.columns
assert "Median (50%)" in statistics_table.columns
assert "Q3 (75%)" in statistics_table.columns

assert not statistics_table.isnull().any().any(), (
    "The statistics table contains missing results."
)

print("All Phase 7 validation tests passed.")

All Phase 7 validation tests passed.


## Task 2.5 — Histograms in a 1×4 Layout

## Task 2.6 — Seaborn Pairplot

## Task 2.7 — Customized Seaborn PairGrid